File: problems/leetcode/853.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        int n = position.size();

        vector<pair<int, int>> cars;
        for (int i = 0; i < n; i++) {
            cars.push_back({position[i], speed[i]});
        }

        sort(cars.begin(), cars.end(), greater<pair<int, int>>());

        stack<double> fleets;

        for (auto& car : cars) {
            double endtime = (double) (target - car.first) / car.second;

            if (fleets.empty() || endtime > fleets.top()) {
                fleets.push(endtime);
            }
        }

        return fleets.size();
    }
};

File: problems/leetcode/853.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        cars = [[pos,spe] for pos, spe in zip(position, speed)]
        fleets = []
        
        for pos, spe in sorted(cars)[::-1]:
            time_taken = (target - pos) / spe
            if not fleets or time_taken > fleets[-1]:
                fleets.append(time_taken)
        
        return len(fleets)